service/controller/machinepoolexp/handler/migration/resource.go

```diff
@@ -0,0 +1,94 @@
+package migration
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v5/service/controller/key"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "migration"
+)
+
+type Config struct {
+	CtrlClient client.Client
+	Logger     micrologger.Logger
+}
+
+type Resource struct {
+	client client.Client
+	logger micrologger.Logger
+}
+
+func New(config Config) (*Resource, error) {
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		client: config.CtrlClient,
+		logger: config.Logger,
+	}
+
+	return r, nil
+}
+
+// EnsureCreated creates non-experimental MachinePool CR if it doesn't already exist.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	// We do this even before upgrade? We must wait for new CRs in the upgrade handler!!!
```

Comment:
what do you mean?

Reply:
A leftover TODO note for myself :smile: 

> We must wait for new CRs in the upgrade handler!!!

I was telling myself that the MachinePool upgrade handler (and all other existing MachinePool handlers) should wait for the migration to be completed, because there are few MachinePool CR updates happening, and existing handlers should wait for the  new MachinePool CR to be fully updated (so they don't work with invalid/incomplete CR).

Non-experimental MachinePool controller (existing controller, but now reconciling new non-experimental MachinePools instead of the old MachinePool) has a new migration handler that is added as the first handler, so it just cancels the CR reconciliation if the migration is not completed. This way all existing handlers will just pick up the fully migrated new non-experimental MachinePool CR and do the same job as before.